### PR TITLE
unsafe_fragment signature change

### PR DESCRIPTION
**Pull request by dallincrane**
Issue https://github.com/elixir-ecto/ecto/issues/2265

### CODE REVIEW

**Comment by dallincrane** [comment]
This explores a `unsafe_fragment(binary(), list())` implementation where `?` characters in the first argument are interpolated with the values from the second argument. Hoping to get some quick feedback as there is a couple different ways this could be done.

I was also thinking it may be nice to have similar function signatures for `fragment` and `unsafe_fragment`. Any thoughts on changing `fragment(...)` to `fragment(binary(), list())` as well? Would that be a bit too frivolous for a breaking change?

File: lib/ecto/query/api.ex [hunk 1]
```diff
@@ -312,7 +312,28 @@ defmodule Ecto.Query.API do
...
+  Sends an unsafe fragment directly to the database.
+
+  WARNING: sending an unsafe fragment may expose your application to sql
+  injection attacks. Think twice before using it and try to reach the same
+  result using the safe fragment/1 API.
+
+  This function will replace any `?` characters in `string` with values
+  from the `fragments` list.
+
+      unsafe_fragment(^"ROW_NUMBER() OVER(ORDER BY ?)", [sort])
```
Comment (ericmj) [inline lib/ecto/query/api.ex]: This example query won't work since a column cannot be a query parameter in SQL.

**Comment by ericmj** [comment]
> Any thoughts on changing fragment(...) to fragment(binary(), list()) as well? Would that be a bit too frivolous for a breaking change?

With fragment we need the list at compile time because that's when we build the query, allowing a list could be confusing because you would think that the following is possible:

```elixir
params = [post.title]
select(query, [post], fragment("?", params))
```

File: lib/ecto/query/builder.ex [hunk 2]
```diff
@@ -82,7 +82,15 @@ defmodule Ecto.Query.Builder do
     {{:{}, [], [:fragment, [], merge_fragments(pieces, frags)]}, params_acc}
   end
 
-  def escape({:unsafe_fragment, _, [{:^, _, [var]} | frags]}, _type, params_acc, vars, env) do
+  def escape({:unsafe_fragment, _, [{:^, _, [var]}]}, _type, params_acc, _vars, _env) do
+    expr = quote do
+      Ecto.Query.Builder.unsafe_fragment!(unquote(var))
```
Comment (josevalim) [inline lib/ecto/query/builder.ex]: We don't need to introduce a new function, just pass an empty list as second argument.

File: lib/ecto/query/builder.ex [hunk 3]
```diff
@@ -583,13 +591,24 @@ defmodule Ecto.Query.Builder do
...
+  Called by escaper at runtime to verify binaries.
+  """
   def unsafe_fragment!(string, fragments) do
     if is_binary(string) do
       pieces = split_fragment(string, "")
 
       if length(pieces) != length(fragments) + 1 do
         raise ArgumentError,
-              "unsafe_fragment(...) expects extra arguments in the same amount of question marks in string"
+              "unsafe_fragment/2 expects to have as many fragment elements as there are question marks in string"
```
Comment (josevalim) [inline lib/ecto/query/builder.ex]: I have seen you renamed `unsafe_fragment(...)` to `unsafe_fragment` almost everywhere except here where you used `unsafe_fragment/2`. Maybe we should use `unsafe_fragment/2` consistently?

**Comment by josevalim** [comment]
@ericmj the problem is that the two APIs are now inconsistent. One uses varargs, the other one does not and uses a list. Maybe we should call this something else rather than unsafe_fragment? I understand though that they are inconsistent exactly because they have different requirements, but I think the differences in API will be confusing.

**Comment by josevalim** [comment]
Hi @dallincrane! After talking to @ericmj we have realized that the old API supported query expressions in interpolation:

    unsafe_fragment("? = ?", p.foo, 1)

But this one does not (or if it does, it is a bit inconsistent as it requires a compile time literal). For this reason, we have decided to keep unsafe_fragment as is.

Maybe we could introduce this PR as a `dynamic_fragment` but I think we should wait until we have more use cases. Thoughts?

**Comment by dallincrane** [comment]
@josevalim, sounds good to me! I would think `#{}` interpolation of the first arg and general string concatenation should work for most cases.

---

I've noticed a couple of differences in the APIs between`fragment(...)` and `unsafe_fragment(...)` that might be good to clear up and I could use this PR to submit some changes if needed!

In the example you gave above I would expect that to work, but the first argument always requires `^`

`from p in "posts", where: unsafe_fragment(^"? = ?", p.foo, 1)`

To summarize

```
fragment("hello") # GOOD
fragment(^"hello") # BAD

unsafe_fragment("hello") # BAD
unsafe_fragment(^"hello") # GOOD
```

I still definitely have a lot to learn here, but it seems like `^` should only be required when the value isn't a compile-time literal. I know that the implementation is a little different when it comes to compile-time vs runtime for fragments. However, as a user, I would still expect the interface to feel similar across the board.

With all 4 examples above being compile-time literals I'm thinking that they should all work?

**Comment by josevalim** [comment]
@dallincrane the thing is that if you are using unsafe_fragment and you have a literal string, then you have no reason to use an unsafe_fragment. :)

**Pull request closed without merging**